### PR TITLE
Credential Scope

### PR DESCRIPTION
**Pull request by dlm6693**
This PR adds a `scope` to credentials that is used for endpoint resolution through the `AWS::AuthCredentialScope` builtin parameter. Similar to account ID, it is treated as completely optional and only set on credential fetchers if the value is present. A couple of properties are added to the endpoint builtin resolver and piped through to the ruleset resolver in order to raise a configuration warning for customers that still rely on legacy behavior.

The base branch is https://github.com/dlm6693/botocore/tree/account-id-endpoints since this PR is largely built on top of that branch and both will be merged at the same time. Once this is approved, account ID will be merged, this branch will be rebased with those changes and the base branch will be updated to https://github.com/boto/botocore/tree/develop/

### CODE REVIEW

File: botocore/regions.py [hunk 1]
```diff
@@ -536,6 +564,22 @@ def _resolve_account_id_builtin(self, param_definitions, builtins):
...
+        scope_builtin_key = EndpointResolverBuiltins.AWS_CREDENTIAL_SCOPE
+        current_builtin_value = builtins.get(scope_builtin_key)
+        credential_resolver = self._resolver_map['credentials']
+        scope = credential_resolver.resolve_credential_scope_builtin(
+            builtin_configured, current_builtin_value
+        )
+        if scope is not None:
+            self._credential_scope_set = True
+        else:
+            self._credential_scope_set = False
```
Comment (dlm6693) [inline botocore/regions.py:580]: While this state is not expected to change between invocations, it needs to be tracked here for the warning.

File: tests/functional/test_endpoint_rulesets.py [hunk 2]
```diff
@@ -305,3 +306,28 @@ def builtin_overwriter_handler(builtins, **kwargs):
...
+@mock.patch(
+    'botocore.regions.EndpointRulesetResolver.auth_schemes_to_signing_ctx',
+    return_value=('v4', {'region': 'us-west-2'}),
+)
+@mock.patch(
+    'botocore.regions.EndpointRulesetResolver.uses_builtin_data_path', False
+)
+@mock.patch(
+    'botocore.regions.EndpointRulesetResolver.credential_scope_set', True
+)
```
Comment (dlm6693) [inline tests/functional/test_endpoint_rulesets.py:321]: It's unfortunate we have to mock so much out, but this is an edge case and would require a ton of scaffolding to be able to handle a warning. We could do this with a mock ruleset instead, but would need to mock out the function that returns the ruleset data to the client. The uses builtin check is already mocked for other behavior [here](https://github.com/boto/botocore/blob/7c1e00cd2bb38a3d41f1ca6849decfbe30605d56/tests/unit/test_args.py#L84) and would be painful to set up. It also seems like too much effort to make a custom endpoints file that would also need to be mocked out.

File: botocore/client.py [hunk 3]
```diff
@@ -1091,13 +1092,34 @@ def _resolve_endpoint_ruleset(
...
 
+    def _maybe_warn_signing_region_mismatch(self, signing_context):
+        legacy_signing_region = self._request_signer.region_name
+        signing_region = signing_context.get('region')
+        if (
+            signing_region is not None
+            and legacy_signing_region != signing_region
+            and not self._ruleset_resolver.uses_builtin_data_path
+            and self._ruleset_resolver.credential_scope_set
+        ):
```
Comment (dlm6693) [inline botocore/client.py:1111]: There's a variety of ways and places we can do this, but I think at the client level makes the most sense because we need state from the endpoint resolver and request signer.  This also avoids adding a new attribute to the signer or adding additional data to the request context for a very niche case.
